File: sourcerer-core/src/main/java/org/elder/sourcerer/ExpectedVersion.java
package org.elder.sourcerer;

import org.elder.sourcerer.exceptions.ConflictingExpectedVersionsException;

/**
 * Describes the version that an aggregate is expected to be in when read, or appended to.
 */
public final class ExpectedVersion {
    private final ExpectedVersionType type;
    private final int expectedVersion;

    private ExpectedVersion(final ExpectedVersionType type, final int expectedVersion) {
        this.type = type;
        this.expectedVersion = expectedVersion;
    }

    /**
     * Gets the type of expected version, see {@link ExpectedVersionType}.
     */
    public ExpectedVersionType getType() {
        return type;
    }

    /**
     * For type == EXACTLY, returns the version that the event stream or aggregate is expected to be
     * in when a command or append is attempted.
     */
    public int getExpectedVersion() {
        return expectedVersion;
    }

    public static ExpectedVersion any() {
        return new ExpectedVersion(ExpectedVersionType.ANY, -1);
    }

    public static ExpectedVersion anyExisting() {
        return new ExpectedVersion(ExpectedVersionType.ANY_EXISTING, -1);
    }

    public static ExpectedVersion notCreated() {
        return new ExpectedVersion(
                ExpectedVersionType.NOT_CREATED,
                Aggregate.VERSION_NOT_CREATED);
    }

    public static ExpectedVersion exactly(final int version) {
        return new ExpectedVersion(ExpectedVersionType.EXACTLY, version);
    }

    public static ExpectedVersion merge(
            final ExpectedVersion baseVersion,
            final ExpectedVersion newVersion) {
        if (baseVersion != null && newVersion != null) {
            if (baseVersion.getType().compareTo(newVersion.getType()) > 0) {
                return mergeOrdered(newVersion, baseVersion);
            } else {
                return mergeOrdered(baseVersion, newVersion);
            }
        } else if (baseVersion != null) {
            return baseVersion;
        } else if (newVersion != null) {
            return newVersion;
        } else {
            return ExpectedVersion.any();
        }
    }

    /**
     * Assumes the expected versions are given in ordinal order.
     */
    private static ExpectedVersion mergeOrdered(
            final ExpectedVersion baseVersion,
            final ExpectedVersion newVersion) {
        switch (baseVersion.getType()) {
            case ANY:
                // Any is always compatible with other expected version
                return newVersion;
            case ANY_EXISTING:
                switch (newVersion.getType()) {
                    case ANY_EXISTING:
                        return ExpectedVersion.anyExisting();
                    case EXACTLY:
                        return newVersion;
                    case NOT_CREATED:
                        throw new ConflictingExpectedVersionsException(
                                "Cannot merge 'any existing' with 'not created'",
                                baseVersion, newVersion);
                    default:
                }
                break;
            case EXACTLY:
                switch (newVersion.getType()) {
                    case EXACTLY:
                        if (baseVersion.getExpectedVersion()
                                != newVersion.getExpectedVersion()) {
                            throw new ConflictingExpectedVersionsException(
                                    "Cannot merge 'exactly "
                                            + baseVersion.getExpectedVersion()
                                            + "' with 'exactly "
                                            + newVersion.getExpectedVersion()
                                            + "'",
                                    baseVersion, newVersion);
                        }
                        return baseVersion;
                    case NOT_CREATED:
                        throw new ConflictingExpectedVersionsException(
                                "Cannot merge 'exactly "
                                        + baseVersion.getExpectedVersion()
                                        + "' with 'not created'",
                                baseVersion, newVersion);
                    default:
                }
                break;
            case NOT_CREATED:
                switch (newVersion.getType()) {
                    case NOT_CREATED:
                        return baseVersion;
                    default:
                }
                break;
            default:
        }
        throw new RuntimeException(
                "Internal error, unexpected version combination merging "
                        + baseVersion
                        + " with " + newVersion);
    }

    @Override
    public String toString() {
        return "ExpectedVersion{" +
                "type=" + type +
                ", expectedVersion=" + expectedVersion +
                '}';
    }
}

File: sourcerer-core/src/main/java/org/elder/sourcerer2/ExpectedVersion.kt
package org.elder.sourcerer2

import org.elder.sourcerer2.exceptions.ConflictingExpectedVersionsException

/**
 * Describes the version that an aggregate is expected to be in when read, or appended to.
 */
sealed class ExpectedVersion {
    data class Exactly(val streamVersion: StreamVersion) : ExpectedVersion()
    object NotCreated : ExpectedVersion() {
        override fun toString(): String {
            return "ExpectedVersion.NotCreated"
        }
    }
    object AnyExisting : ExpectedVersion()  {
        override fun toString(): String {
            return "ExpectedVersion.AnyExisting"
        }
    }
    object Any : ExpectedVersion()  {
        override fun toString(): String {
            return "ExpectedVersion.Any"
        }
    }

    companion object {
        // Java convenience methods.
        @JvmStatic
        fun any(): ExpectedVersion {
            return ExpectedVersion.Any
        }

        @JvmStatic
        fun anyExisting(): ExpectedVersion {
            return ExpectedVersion.AnyExisting
        }

        @JvmStatic
        fun notCreated(): ExpectedVersion {
            return ExpectedVersion.NotCreated
        }

        @JvmStatic
        fun exactly(version: StreamVersion): ExpectedVersion {
            return ExpectedVersion.Exactly(version)
        }

        @JvmStatic
        fun merge(
                baseVersion: ExpectedVersion?,
                newVersion: ExpectedVersion?
        ): ExpectedVersion {
            return if (baseVersion != null && newVersion != null) {
                if (baseVersion.ordinal() > newVersion.ordinal()) {
                    mergeOrdered(newVersion, baseVersion)
                } else {
                    mergeOrdered(baseVersion, newVersion)
                }
            } else baseVersion ?: (newVersion ?: ExpectedVersion.any())
        }

        /**
         * Assumes the expected versions are given in ordinal order.
         */
        private fun mergeOrdered(
                baseVersion: ExpectedVersion,
                newVersion: ExpectedVersion
        ): ExpectedVersion {
            return when (baseVersion) {
                ExpectedVersion.Any ->
                    // Any is always compatible with other expected version
                    newVersion
                ExpectedVersion.AnyExisting -> when (newVersion) {
                    ExpectedVersion.AnyExisting,
                    ExpectedVersion.Any -> ExpectedVersion.anyExisting()
                    ExpectedVersion.NotCreated -> throw ConflictingExpectedVersionsException(
                            "Cannot merge 'any existing' with 'not created'",
                            baseVersion, newVersion)
                    is ExpectedVersion.Exactly -> newVersion
                }
                is ExpectedVersion.Exactly -> when (newVersion) {
                    is ExpectedVersion.Exactly -> {
                        if (baseVersion.streamVersion != newVersion.streamVersion) {
                            throw ConflictingExpectedVersionsException(
                                    "Cannot merge 'exactly "
                                            + baseVersion.streamVersion
                                            + "' with 'exactly "
                                            + newVersion.streamVersion
                                            + "'",
                                    baseVersion, newVersion)
                        }
                        baseVersion
                    }
                    ExpectedVersion.NotCreated ->
                        throw ConflictingExpectedVersionsException(
                                "Cannot merge 'exactly "
                                        + baseVersion.streamVersion
                                        + "' with 'not created'",
                                baseVersion, newVersion)
                    else ->
                        throw RuntimeException(
                                "Internal error, unexpected version combination merging "
                                        + baseVersion
                                        + " with " + newVersion)

                }
                ExpectedVersion.NotCreated -> when (newVersion) {
                    ExpectedVersion.NotCreated -> baseVersion
                    else ->
                        throw RuntimeException(
                                "Internal error, unexpected version combination merging "
                                        + baseVersion
                                        + " with " + newVersion)
                }
            }
        }
    }
}

private fun ExpectedVersion.ordinal(): Int {
    return when (this) {
        ExpectedVersion.Any -> 0
        ExpectedVersion.AnyExisting -> 1
        is ExpectedVersion.Exactly -> 2
        ExpectedVersion.NotCreated -> 3
    }
}
